lib/elixir/src/elixir_erl.erl

```diff
@@ -559,6 +581,38 @@ get_type_docs(Set, Types) ->
      {_, Line, Doc} <- ets:lookup(Set, {Kind, Tuple})]
   ).
 
+translate_doc({{Name, Arity}, Line, Kind, Signature, Doc}) ->
+  {{translate_doc_kind(Kind), Name, Arity},
+   erl_anno:new(Line),
+   [signature_to_binary(Name, Signature)],
+   doc_value(Doc),
+   #{signature => Signature}
+  };
+translate_doc({{Name, Arity}, Line, Kind, Doc}) ->
+  {{translate_doc_kind(Kind), Name, Arity},
+   erl_anno:new(Line),
+   [],
+   doc_value(Doc),
+   #{}
+  }.
+
+translate_doc_kind(def)           -> function;
+translate_doc_kind(callback)      -> callback;
+translate_doc_kind(type)          -> type;
+translate_doc_kind(opaque)        -> opaque;
+translate_doc_kind(defmacro)      -> macro;
+translate_doc_kind(macrocallback) -> macrocallback.
+
+signature_to_binary(Name, Signature) ->
```

Comment:
I believe we should be able to replace this whole function by `'Elixir.Macro':to_string({Name, [], Signature})`. Thoughts?

Reply:
Hmm, that works. Why isn't IEx doing the same?

Reply:
Oh... no idea. :)

Reply:
The change for this made CI fail and I could reproduce it locally by running `make clean` first. The error is the following:

```
== Compilation error in file lib/kernel/special_forms.ex ==
** (MatchError) no match of right hand side value: {:map, [], nil}
    /Users/LacKac/Code/elixir-lang/elixir/lib/elixir/lib/macro.ex:623: Macro.to_string/2
    (stdlib) erl_eval.erl:677: :erl_eval.do_apply/6
** (UndefinedFunctionError) function List.Chars.to_charlist/1 is undefined (module List.Chars is not available)
    List.Chars.to_charlist("lib/kernel/special_forms.ex")
    /Users/LacKac/Code/elixir-lang/elixir/lib/elixir/lib/kernel/parallel_compiler.ex:500: Kernel.ParallelCompiler.get_line/3
    /Users/LacKac/Code/elixir-lang/elixir/lib/elixir/lib/kernel/parallel_compiler.ex:483: Kernel.ParallelCompiler.to_error/4
    /Users/LacKac/Code/elixir-lang/elixir/lib/elixir/lib/kernel/parallel_compiler.ex:383: Kernel.ParallelCompiler.wait_for_messages/6
    /Users/LacKac/Code/elixir-lang/elixir/lib/elixir/lib/kernel/parallel_compiler.ex:135: Kernel.ParallelCompiler.spawn_workers/3
    /Users/LacKac/Code/elixir-lang/elixir/lib/elixir/lib/kernel/cli.ex:469: anonymous fn/2 in Kernel.CLI.process_command/2
make: *** [lib/elixir/ebin/Elixir.Kernel.beam] Error 1
```

Reply:
@lackac it is a bootstrap issue. We need some modules to compile Elixir itself and this change makes List.Chars required. Can you add `<<"lib/elixir/lib/list/chars.ex">>` [below here](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/src/elixir_compiler.erl#L151) and it should work. I will investigate after merging to remove back this dependency.

Reply:
@josevalim protip, when you hit `y` in diff then URL will be expanded to canonical form with `master` replaced with commit SHA.

Reply:
@josevalim it still fails to compile:

```
== Compilation error in file lib/kernel/special_forms.ex ==
** (MatchError) no match of right hand side value: {:map, [], nil}
    /Users/LacKac/Code/elixir-lang/elixir/lib/elixir/lib/macro.ex:623: Macro.to_string/2
    (stdlib) erl_eval.erl:677: :erl_eval.do_apply/6
make: *** [lib/elixir/ebin/Elixir.Kernel.beam] Error 1
```

I think it's because it's trying to generate the signature for the `%Struct{}` special form. The signature AST is `{:%, [], [{:struct, [], nil}, {:map, [], nil}]}` which `Macro.to_string/2` is trying to render as a `%StructName{}` expression, but the right hand side is not a map representation, so it fails on that.

If I add a special case just for this special form it successfully compiles the rest even without adding `List.Chars` to the bootstrap list. After doing that I wanted to see what signatures were generated for the special form macros. Here's the list:

``` elixir
iex(4)> for {{:macro, n, a}, _, [s], _, _} <- docs, do: {{n, a}, s}
[
  {{:%, 2}, "%struct_name{}"},
  {{:%{}, 1}, "%{args}"},
  {{:&, 1}, "&(expr)"},
  {{:., 2}, "left . right"},
  {{:::, 2}, "left :: right"},
  {{:<<>>, 1}, "<<args>>"},
  {{:=, 2}, "left = right"},
  {{:^, 1}, "^var"},
  {{:__CALLER__, 0}, "__CALLER__()"},
  {{:__DIR__, 0}, "__DIR__()"},
  {{:__ENV__, 0}, "__ENV__()"},
  {{:__MODULE__, 0}, "__MODULE__()"},
  {{:__STACKTRACE__, 0}, "__STACKTRACE__()"},
  {{:__aliases__, 1}, "args"},
  {{:__block__, 1}, "args"},
  {{:alias, 2}, "alias(module, opts)"},
  {{:case, 2}, "case(condition, clauses)"},
  {{:cond, 1}, "cond(clauses)"},
  {{:fn, 1}, "fn\n  [clauses]\nend"},
  {{:for, 1}, "for(args)"},
  {{:import, 2}, "import(module, opts)"},
  {{:quote, 2}, "quote(opts, block)"},
  {{:receive, 1}, "receive(args)"},
  {{:require, 2}, "require(module, opts)"},
  {{:super, 1}, "super(args)"},
  {{:try, 1}, "try(args)"},
  {{:unquote, 1}, "unquote(expr)"},
  {{:unquote_splicing, 1}, "unquote_splicing(expr)"},
  {{:with, 1}, "with(args)"},
  {{:{}, 1}, "{args}"}
]
```

As you can see some of these don't make much sense. It's also not much better than what the implementation borrowed from IEx was doing.

I see three options:

1. keep the original naive implementation based on IEx
2. use `'Elixir.Macro':to_string/2` but with a special case for `%struct_name{}`
3. generalise on the special case and allow macros to define their own signature (e.g. `@signature "%struct_name{key: value, ...}"`)

Reply:
I see. Let's keep the original naive implementation based on IEx then. :)